src/celutil/tokenizer.cpp

```diff
@@ -8,622 +8,800 @@
-    NumberSigned,
-    Number,
+    Initial,
     Fraction,
-    ExponentStart,
-    ExponentSigned,
     Exponent,
-    Name,
-    String,
-    StringEscape,
-    UnicodeEscape,
-    Comment,
+    End,
+    Error,
 };
 
 
-bool isSeparator(unsigned char u)
+struct NumberState
 {
-    constexpr auto period = static_cast<unsigned char>('.');
-    return !std::isdigit(u) && !std::isalpha(u) && u != period;
-}
+    std::size_t endPosition{ 0 };
+    NumberPart part{ NumberPart::Initial };
+    bool isInteger{ true };
+};
 
 
-bool tryPushBack(std::string& s, char c)
+struct StringState
 {
-    if (s.size() < maxTokenLength)
-    {
-        s.push_back(c);
-        return true;
-    }
+    std::size_t runStart{ 1 };
+    std::size_t runEnd{ 1 };
+    std::string processed{};
+    UTF8Validator validator{};
 
-    GetLogger()->error("Token too long\n");
-    return false;
-}
+    bool checkUTF8(char, std::string_view);
+};
 
 
-bool handleUtf8Error(std::string& s, UTF8Status status)
+bool
+StringState::checkUTF8(char c, std::string_view run)
 {
-    if (status == UTF8Status::InvalidTrailingByte)
+    auto utf8Status = validator.check(c);
+    if (utf8Status == UTF8Status::Ok) { return true; }
+
+    if (utf8Status == UTF8Status::InvalidTrailingByte)
     {
-        // remove the partial UTF-8 sequence
-        std::string::size_type pos = s.size();
+        std::size_t pos = run.size();
         while (pos > 0)
         {
-            unsigned char u = static_cast<unsigned char>(s[--pos]);
-            if (u >= 0xc0) { break; }
+            --pos;
+            auto uch = static_cast<unsigned char>(run[pos]);
+            if (uch >= 0xc0) { break; }
         }
 
-        s.resize(pos);
+        run = run.substr(0, pos);
     }
 
-    if (s.size() <= maxTokenLength - std::strlen(UTF8_REPLACEMENT_CHAR))
-    {
-        s.append(UTF8_REPLACEMENT_CHAR);
-        return true;
-    }
-
-    GetLogger()->error("Token too long\n");
+    processed.append(run);
+    processed.append(UTF8_REPLACEMENT_CHAR);
+    runStart = runEnd + 1;
+    runEnd = runStart;
     return false;
 }
+
+
+struct VisitorStringView
+{
+    std::string_view operator()(std::monostate) const { return {}; }
+    std::string_view operator()(std::int32_t) const { return {}; }
+    std::string_view operator()(double) const { return {}; }
+    std::string_view operator()(std::string_view sv) const { return sv; }
+    std::string_view operator()(const std::string& s) const { return s; }
+};
+
+
+struct VisitorDouble
+{
+    double operator()(std::monostate) const { return std::nan(""); }
+    double operator()(std::int32_t i) const { return static_cast<double>(i); }
+    double operator()(double d) const { return d; }
+    double operator()(std::string_view) const { return std::nan(""); }
+    double operator()(const std::string&) const { return std::nan(""); }
+};
+
+
+struct VisitorInteger
+{
+    std::optional<std::int32_t> operator()(std::monostate) const { return std::nullopt; }
```

Comment:
is it a good idea to provide inconsistent api for the similar purpose? option for integers and actual types for double/string_view?

Reply:
Probably not, I've updated this, plus separated out `getNameValue()` and `getStringValue()` to combine the token type tests and data retrieval